### PR TITLE
Roundend report refactor

### PR DESCRIPTION
**Pull request by AnturK**
dreamseeker_2017-12-03_20-19-59

TODO:
- [x] Cleanup/Debug removal/Moving things to proper files
- [x] Reimplement feedback
- [x] Unify classses used for styling these
- [ ] Testing
- [ ] Actually implement the remaining mini-antags datums.

### CODE REVIEW

File: code/__HELPERS/roundend.dm [hunk 1]
```diff
@@ -0,0 +1,303 @@
+/datum/controller/subsystem/ticker/proc/gather_roundend_feedback()
+	var/clients = 0
+	var/surviving_humans = 0
+	var/surviving_total = 0
+	var/ghosts = 0
+	var/escaped_humans = 0
+	var/escaped_total = 0
+
+	for(var/mob/M in GLOB.player_list)
+		if(M.client)
```
Comment (Cyberboss) [inline code/__HELPERS/roundend.dm]: redundant check

File: code/__HELPERS/roundend.dm [hunk 2]
```diff
@@ -0,0 +1,303 @@
...
+			clients++
```
Comment (Cyberboss) [inline code/__HELPERS/roundend.dm]: this can just be GLOB.player_list.len

**Review by Cyberboss — DISMISSED** [review]

**Comment by Jordie0608** [comment]
#33292?

I think your two new feedbacks would be better as associative types but I'll have to check how nested lists end up encoded in them.

**Comment by AnturK** [comment]
Yeah, i thought it would be simple enough like this but i forgot about objectiveless antags, and i'm not really happy with the format.

**Comment by Jordie0608** [comment]
To get lists in the associative type working replace:
https://github.com/tgstation/tgstation/blob/9aa8a911fa963ac2f40d393b1f0384ae80cb2265/code/controllers/subsystem/blackbox.dm#L212-L213
with:
~~~
for(var/i in data)
	if(islist(data[i]))
		FV.json["data"]["[pos]"]["[i]"] = data[i] //and here with "[FV.json["data"].len]"
	else
		FV.json["data"]["[pos]"]["[i]"] = "[data[i]]" //and here with "[FV.json["data"].len]"
~~~
Tested with a 3 deep list, seemed to work.

**Comment by nfreader** [comment]
Can feedback get a "clean" antagonist type? `/datum/antagonist/traitor/human` is good, but it'd be a lot cleaner if there was a field that just said something like `Traitor` or `Head Revolutionary`.

**Comment by AnturK** [comment]
There is one already ?

File: code/datums/mind.dm [hunk 3]
```diff
@@ -769,17 +762,44 @@
...
 				def_value = "custom"
+		else
+			switch(antag_datums.len)
+				if(0)
+					target_antag = add_antag_datum(/datum/antagonist/custom)
+				if(1)
+					target_antag = antag_datums[1]
+				else
+					var/target = input("Which antagonist gets the objective:", "Antagonist", def_value) as null|anything in antag_datums + "(new custom antag)"
+					if (!target)
```
Comment (optimumtact) [inline code/datums/mind.dm]: !QDELETED(target)

File: code/game/machinery/wishgranter.dm [hunk 4]
```diff
@@ -41,10 +41,11 @@
 
 		SSticker.mode.traitors += user.mind
 		user.mind.special_role = "Avatar of the Wish Granter"
-
+
```
Comment (optimumtact) [inline code/game/machinery/wishgranter.dm]: stray space

**Review by optimumtact — DISMISSED** [review]
Minor nits, but this looks good to me

File: code/datums/mind.dm [hunk 5]
```diff
@@ -769,17 +762,44 @@
...
 				def_value = "custom"
+		else
+			switch(antag_datums.len)
+				if(0)
+					target_antag = add_antag_datum(/datum/antagonist/custom)
+				if(1)
+					target_antag = antag_datums[1]
+				else
+					var/target = input("Which antagonist gets the objective:", "Antagonist", def_value) as null|anything in antag_datums + "(new custom antag)"
+					if (QDELETED(target))
```
Comment (Jordie0608) [inline code/datums/mind.dm]: `code/datums/mind.dm:797:error: target.gc_destroyed: undefined var`

Comment (optimumtact) [reply code/datums/mind.dm]: ugh yeah, you might need to type to var/datum/target

**Comment by AnturK** [comment]
New css from @vuonojenmustaturska :

**Comment by MoreRobustThanYou** [comment]
@AnturK I like that CSS, although i'd like the red to be changed a bit.

File: code/__HELPERS/roundend.dm [hunk 6]
```diff
@@ -0,0 +1,412 @@
...
+
+	gather_antag_success_rate()
+
+/datum/controller/subsystem/ticker/proc/gather_antag_success_rate()
+	var/team_gid = 1
+	var/list/team_ids = list()
+
+	for(var/datum/antagonist/A in GLOB.antagonists)
+		var/list/antag_info = list()
+		antag_info["ckey"] = A.owner.key
```
Comment (Cyberboss) [inline code/__HELPERS/roundend.dm]: one of these things is not like the other?

Comment (AnturK) [reply code/__HELPERS/roundend.dm]: You mean the key/ckey or what ? Minds should be changed to use ckey instead of key but that's separate issue.

**Comment by Cyberboss** [comment]
What_could_go_wrong.commit

**Pull request merged**